File: app/src/main/java/cn/studyjams/s220170131/mijack/core/MediaManager.java
package cn.studyjams.s220170131.mijack.core;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Log;

import cn.studyjams.s220170131.mijack.entity.Folder;
import cn.studyjams.s220170131.mijack.entity.Image;
import cn.studyjams.s220170131.mijack.entity.Media;
import cn.studyjams.s220170131.mijack.util.Utils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Mr.Yuan
 * @date 2017/4/17
 */
public class MediaManager {
    public static final String TAG = "MediaManager";

    public static List<Folder> getImageFolderWithImages(Context context) {
        return getImageFolderWithImages(context, -1);
    }

    public static List<Folder> getImageFolderWithImages(Context context, int limit) {
        ContentResolver contentResolver = context.getContentResolver();
        List<Folder> result = new ArrayList<>();
        //table images
        //table files
        Uri contentUri = MediaStore.Files.getContentUri("external");
        Cursor folderCursor = contentResolver.query(contentUri
                , new String[]{MediaStore.Images.ImageColumns.DATA, "count(*)", MediaStore.Files.FileColumns.PARENT},
                MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE
                        + ") group by " +
                        MediaStore.Files.FileColumns.PARENT + " --"
                , null, null);
        if (folderCursor != null && folderCursor.moveToFirst()) {
            do {
                System.out.println("-----------------------------------------------------");
                File file = new File(folderCursor.getString(0));
                int count = folderCursor.getInt(1);
                int parentId = folderCursor.getInt(2);
                List<Image> images = getImagesFromFolder(context, file.getParent(), limit);
                Log.d(TAG, "getImageFoldersFromMedia: size:" + Utils.size(images));
                Log.d(TAG, "getImageFoldersFromMedia: count:" + count);
                Log.d(TAG, "getImageFoldersFromMedia: parent:" + parentId);
                Folder imageMediaFolder = new Folder(file.getParent());
                imageMediaFolder.setData(images);
                imageMediaFolder.setCount(count);
                result.add(imageMediaFolder);
            } while (folderCursor.moveToNext());
        }

        Utils.close(folderCursor);
        return result;
    }

    private static List<Image> getImagesFromFolder(Context context, String folder, int limit) {
        List<Image> images = new ArrayList<>();
        ContentResolver contentResolver = context.getContentResolver();
        Uri contentUri = MediaStore.Files.getContentUri("external");
        String select = MediaStore.Images.Media.DATA + " like '" + folder + "%' and "
                + MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE
                + ") limit " + limit + /*" order by" + MediaStore.Files.FileColumns.DATE_MODIFIED +*/ " --";
        System.out.println(select);
        Cursor c = contentResolver.query(contentUri, new String[]{
                MediaStore.Images.Media._ID,
                MediaStore.Images.Media.DATA,
                MediaStore.Images.Media.SIZE,
                MediaStore.Images.Media.DISPLAY_NAME,
                MediaStore.Images.Media.DATE_TAKEN,
                MediaStore.Images.Media.MIME_TYPE,
                MediaStore.Images.Media.WIDTH,
                MediaStore.Images.Media.HEIGHT,
                MediaStore.Images.Media.LATITUDE,
                MediaStore.Images.Media.LONGITUDE,
                MediaStore.Images.Media.ORIENTATION
        }, select, null, null);
        if (c != null && c.moveToFirst()) {
            Log.d(TAG, "getImagesFromFolder: cursor count:" + c.getCount());
            do {
                System.out.println(
                        c.getString(c.getColumnIndex(MediaStore.Images.Media.DATA)));
                Image image = Image.getImageFromCursor(c);
                images.add(image);
            } while (c.moveToNext());
        }
        Utils.close(c);
        return images;
    }


    public static List<Media> flatFolder(List<Folder> folders) {
        List<Media> result = new ArrayList<>();
        for (Folder folder : folders) {
            result.add(folder);
            result.addAll(folder.getData());
        }
        return result;
    }

    public static List<Image> getImages(Context context) {
        List<Image> images = new ArrayList<>();
        ContentResolver contentResolver = context.getContentResolver();
        Uri contentUri = MediaStore.Files.getContentUri("external");
        String select = MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE;
        System.out.println(select);
        Cursor c = contentResolver.query(contentUri, new String[]{
                MediaStore.Images.Media._ID,
                MediaStore.Images.Media.DATA,
                MediaStore.Images.Media.SIZE,
                MediaStore.Images.Media.DISPLAY_NAME,
                MediaStore.Images.Media.DATE_TAKEN,
                MediaStore.Images.Media.MIME_TYPE,
                MediaStore.Images.Media.WIDTH,
                MediaStore.Images.Media.HEIGHT,
                MediaStore.Images.Media.LATITUDE,
                MediaStore.Images.Media.LONGITUDE,
                MediaStore.Images.Media.ORIENTATION
        }, select, null, null);
        if (c != null && c.moveToFirst()) {
            Log.d(TAG, "getImages: cursor count:" + c.getCount());
            do {
                System.out.println(
                        c.getString(c.getColumnIndex(MediaStore.Images.Media.DATA)));
                Image image = Image.getImageFromCursor(c);
                images.add(image);
            } while (c.moveToNext());
        }
        Utils.close(c);
        return images;
    }
}

File: app/src/main/java/cn/mijack/imagedrive/core/MediaManager.kt
package cn.mijack.imagedrive.core

import android.content.Context
import android.provider.MediaStore
import android.support.media.ExifInterface
import cn.mijack.imagedrive.entity.Folder
import cn.mijack.imagedrive.entity.Image
import cn.mijack.imagedrive.entity.Media
import cn.mijack.imagedrive.util.Utils
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException

/**
 * @author admin
 * @date 2017/8/30
 */
class MediaManager {

    companion object {
        fun getImageFolderWithImages(context: Context): List<Folder> {
            return getImagesFromFolder(context, -1)
        }

        fun getImagesFromFolder(context: Context, limit: Int): List<Folder> {
            var contentResolver = context.contentResolver
            var result = ArrayList<Folder>()
            //table images
            //table files
            var contentUri = MediaStore.Files.getContentUri("external")
            var folderCursor = contentResolver.query(contentUri
                    , Array<String>(3) { MediaStore.Images.ImageColumns.DATA; "count(*)"; MediaStore.Files.FileColumns.PARENT },
                    MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE
                            + ") group by " +
                            MediaStore.Files.FileColumns.PARENT + " --"
                    , null, null)
            if (folderCursor != null && folderCursor.moveToFirst()) {
                do {
                    System.out.println("-----------------------------------------------------")
                    var file = File(folderCursor.getString(0))
                    var count = folderCursor.getInt(1)
                    var parentId = folderCursor.getInt(2)
                    var images = getImagesFromFolder(context, file.getParent(), limit)
                    var imageMediaFolder = Folder(file.getParent())
                    imageMediaFolder.data = images//(images)
                    imageMediaFolder.count = count
                    result.add(imageMediaFolder)
                } while (folderCursor.moveToNext())
            }

            Utils.close(folderCursor)
            return result
        }

        fun getImagesFromFolder(context: Context, folder: String, limit: Int): List<Image> {
            var images = ArrayList<Image>()
            var contentResolver = context.contentResolver
            var contentUri = MediaStore.Files.getContentUri("external")
            var select = MediaStore.Images.Media.DATA +
                    " like '" + folder + "%' and " +
                    MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE + ") " +
                    (if (limit > 0) ("limit " + limit) else "") + " --"
            var c = contentResolver.query(contentUri, Array(11) {
                MediaStore.Images.Media._ID; MediaStore.Images.Media.DATA; MediaStore.Images.Media.SIZE;
                MediaStore.Images.Media.DISPLAY_NAME; MediaStore.Images.Media.DATE_TAKEN; MediaStore.Images.Media.MIME_TYPE;
                MediaStore.Images.Media.WIDTH; MediaStore.Images.Media.HEIGHT; MediaStore.Images.Media.LATITUDE;
                MediaStore.Images.Media.LONGITUDE; MediaStore.Images.Media.ORIENTATION
            }, select, null, null)
            if (c != null && c.moveToFirst()) {
                do {
                    System.out.println(
                            c.getString(c.getColumnIndex(MediaStore.Images.Media.DATA)));
                    var image = Image.getImageFromCursor(c)
                    images.add(image)
                } while (c.moveToNext())
            }
            Utils.Companion.close(c)
            return images
        }

        fun flatFolder(folders: List<Folder>): List<Media> {
            var result = ArrayList<Media>()
            for (f: Folder in folders) {
                result.add(f)
                result.addAll(f.data)
            }
            return result
        }

        fun getImagesInFolder(context: Context, folderPath: String): List<Image> {
            return getImagesFromFolder(context, folderPath, -1)
        }

        fun getImages(context: Context): List<Image> {
            var images = ArrayList<Image>();
            var contentUri = MediaStore.Files.getContentUri("external");
            var select = MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE;
            var c = context.contentResolver.query(contentUri, Array<String>(11) {
                MediaStore.Images.Media._ID;
                MediaStore.Images.Media.DATA;
                MediaStore.Images.Media.SIZE;
                MediaStore.Images.Media.DISPLAY_NAME;
                MediaStore.Images.Media.DATE_TAKEN;
                MediaStore.Images.Media.MIME_TYPE;
                MediaStore.Images.Media.WIDTH;
                MediaStore.Images.Media.HEIGHT;
                MediaStore.Images.Media.LATITUDE;
                MediaStore.Images.Media.LONGITUDE;
                MediaStore.Images.Media.ORIENTATION
            }, select, null, null)
            if (c != null && c.moveToFirst()) {
                do {
                    System.out.println(
                            c.getString(c.getColumnIndex(MediaStore.Images.Media.DATA)))
                    var image = Image.getImageFromCursor(c)
                    images.add(image)
                } while (c.moveToNext())
            }
            Utils.close(c)
            return images;
        }

        //
        fun getExifInterface(imagePath: String): ExifInterface? {
            try {
                var fis = FileInputStream(imagePath)
                var exifInterface = ExifInterface(fis)
//            exifInterface.
                return exifInterface
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return null
        }

        //
        fun deleteFile(path: String): Boolean {
            var file = File(path)
            return file.delete()
        }
    }
//
}

